dataplane/api-server/src/config.rs

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+*/
+use anyhow::{anyhow, Result};
+use clap::Parser;
+use std::path::PathBuf;
+
+#[derive(Debug, Parser, Clone)]
+pub struct TLSConfig {
+    #[clap(long, default_value = "false")]
+    pub enable_tls: bool,
+    #[clap(long, default_value = "false")]
+    pub enable_mtls: bool,
+    pub server_certificate_path: Option<PathBuf>,
+    pub server_private_key_path: Option<PathBuf>,
+    pub client_certificate_authority_root_path: Option<PathBuf>,
+}
+
+impl TLSConfig {
+    pub fn validate(&self) -> Result<()> {
+        let tls_paths = &[
+            (&self.server_certificate_path, "server_certificate_path"),
+            (&self.server_private_key_path, "server_private_key_path"),
+        ];
+
+        let mtls_paths = &[(
+            &self.client_certificate_authority_root_path,
+            "client_certificate_authority_root_path",
+        )];
+
+        if self.enable_mtls {
+            validate_paths(tls_paths)?;
+            validate_paths(mtls_paths)?;
+        } else if self.enable_tls {
+            validate_paths(tls_paths)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn validate_paths(paths: &[(&Option<PathBuf>, &str)]) -> Result<()> {
+    for (path, name) in paths {
+        if path.is_none() {
+            return Err(anyhow!("Missing required path: {}", name));
+        }
+    }
+    Ok(())
+}
```

Comment:
If we go with the `Enum` approach outlined above, then none of this Validation would be needed, but we could create [associated functions](https://doc.rust-lang.org/stable/book/ch05-03-method-syntax.html#associated-functions) to [construct](https://rust-unofficial.github.io/patterns/idioms/ctor.html) these variants.

Reply:
Is this even necessary? Shouldn't clap handle this for us during parsing? The only reason I was doing this was to enforce the conditional dependencies between fields. The enum renders that irrelevant.

Can't we just pass `opt.tls_config` down to the `setup_tls` function and match on the variant?

Reply:
Not necessary no, just a thought depending on how you'd like this to end up looking. But the spirit behind it was in-line with the second thing you were suggesting: once we use an enum "validation" does effectively just get outmoded by matching on the variants and applying the appropriate functionality thereafter. :+1: